File: Modeller_Section/modeller10.4/examples/c-extensions/cuser_optimizer.c
#include <glib.h>
#include <math.h>
#include <stdlib.h>
#include "modeller.h"
#include "cuser_optimizer.h"

/** Main optimization loop */
void mainloop(struct mod_state_optimizer *opt, struct mod_energy_data *edat,
              const struct mod_libraries *libs, float alpha, float minshift,
              float min_ediff, int maxit, int *ierr)
{
  float *state, *dstate, olde;
  int n_state, n_dstate;

  *ierr = 0;

  mod_state_optimizer_state_get(opt, &state, &n_state);
  mod_state_optimizer_energy(opt, edat, state, n_state, &olde, &dstate,
                             &n_dstate, libs, ierr);
  if (*ierr != 0) {
    g_free(state);
    return;
  }

  while (*ierr == 0) {
    int i;
    float newe;

    for (i = 0; i < n_state; i++) {
      state[i] -= alpha * dstate[i];
    }
    g_free(dstate);
    mod_state_optimizer_energy(opt, edat, state, n_state, &newe, &dstate,
                               &n_dstate, libs, ierr);
    if (*ierr != 0) {
      g_free(state);
      return;
    }

    if (fabs(newe - olde) < min_ediff) {
      mod_logout("Finished at step %d due to energy criterion", opt->opt.step);
      break;
    } else if (opt->opt.shiftmax < minshift) {
      mod_logout("Finished at step %d due to shift criterion", opt->opt.step);
      break;
    } else if (opt->opt.step >= maxit) {
      mod_logout("Finished at step %d due to step criterion", opt->opt.step);
      break;
    }
    if (newe < olde) {
      alpha *= 2.0;
    } else {
      alpha /= 2.0;
    }
    olde = newe;
    mod_state_optimizer_next_step(opt, ierr);
  }
  mod_state_optimizer_finish(opt, ierr);
  g_free(state);
  g_free(dstate);
}

File: Modeller_Section/modeller10.4/examples/c-extensions/cuser_optimizer.py
from modeller.optimizers import StateOptimizer

# Load the C extension module; this needs to be compiled first - see
# cuser_feat.py for suitable commands.
import _cuser_optimizer

class SteepestDescent(StateOptimizer):
    """Very simple steepest descent optimizer, in C"""

    # Add options for our optimizer
    _ok_keys = StateOptimizer._ok_keys + ('min_atom_shift', 'min_e_diff',
                                          'step_size', 'max_iterations')

    def __init__(self, step_size=0.0001, min_atom_shift=0.01, min_e_diff=1.0,
                 max_iterations=None, **vars):
        StateOptimizer.__init__(self, step_size=step_size,
                                min_atom_shift=min_atom_shift,
                                min_e_diff=min_e_diff,
                                max_iterations=max_iterations, **vars)

    def optimize(self, atmsel, **vars):
        # Do normal optimization startup
        StateOptimizer.optimize(self, atmsel, **vars)

        # Get all parameters
        alpha = self.get_parameter('step_size')
        minshift = self.get_parameter('min_atom_shift')
        min_ediff = self.get_parameter('min_e_diff')
        maxit = self.get_parameter('max_iterations')

        (opt, edat, libs) = self.get_modeller_objects()
        ierr = _cuser_optimizer.mainloop(opt, edat, libs, alpha, minshift,
                                         min_ediff, maxit)
        if ierr != 0:
            raise ModellerError
